tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CRT_SECURE_NO_WARNINGS
+#include <algorithm>
+#include <array>
+#include <assert.h>
+#include <exception>
+#include <filesystem>
+#include <stdexcept>
+#include <stdio.h>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+using namespace std;
+
+class BinaryFile {
+public:
+    explicit BinaryFile(const string& filepath) {
+        m_file = fopen(filepath.c_str(), "rb");
+
+        if (!m_file) {
+            throw runtime_error("fopen() failed for: " + filepath);
+        }
+    }
+
+    [[nodiscard]] bool read_next_block(vector<unsigned char>& buffer) {
+        constexpr size_t BlockSize = 65536;
+
+        buffer.resize(BlockSize);
+
+        const size_t bytes_read = fread(buffer.data(), 1, BlockSize, m_file);
+
+        buffer.resize(bytes_read);
+
+        return !buffer.empty();
+    }
+
+    ~BinaryFile() {
+        if (fclose(m_file) != 0) {
+            fprintf(stderr, "fclose() failed.\n");
+            terminate();
+        }
+    }
+
+    BinaryFile(const BinaryFile&) = delete;
+    BinaryFile& operator=(const BinaryFile&) = delete;
+
+private:
+    FILE* m_file{nullptr};
+};
+
+void scan_file(const string& filepath, vector<unsigned char>& buffer) {
+    constexpr char CR = '\r';

```

Comment:
Should we use `char8_t` and `u8` character literals in this function as well?

Reply:
As mentioned in my other reply, since one of the purposes of this validator is to disallow UTF-8 in general, I believe that `char8_t` and `u8` character literals would serve no purpose here. I would rather replace these character constants with hex constants if you'd like to avoid any `char` interaction at all.